pymarketstore/results.py

```diff
@@ -67,88 +66,97 @@ def df(self):
+        DataFrame indexed by symbol with the queried attribute groups as the
+        columns (eg OHLCV -> (Open, High, Low, Close, Volume))
+        """
+        timeframe = timeframe or self._default_timeframe
+        columns = self.first().array.dtype.names
+        symbols = []
+        latest_bars = []
+
+        for ds in self._timeframes[timeframe].values():
+            symbols.append(ds.symbol)
+            latest_bars.append(ds.array[-1].tolist())
+
+        df = pd.DataFrame(latest_bars, index=symbols, columns=columns)
+        if 'Epoch' in columns:
+            df.Epoch = pd.to_datetime(df.Epoch, unit='s', utc=True)
+        return df
 
-class QueryReply(object):
+    def keys(self):
+        return list(self._datasets.keys())
 
-    def __init__(self, reply):
-        results = decode_responses(reply['responses'])
-        self.results = [QueryResult(result, reply) for result in results]
-        self.reply = reply
+    def symbols(self):
+        return list(self._symbols.keys())
 
-    @property
-    def timezone(self):
-        return self.reply['timezone']
+    def by_symbols(self, timeframe=None):
+        timeframe = timeframe or self._default_timeframe
+        return {symbol: data[timeframe]
+                for symbol, data in six.iteritems(self._symbols)}
 
-    def first(self):
-        return self.results[0].first()
+    def timeframes(self):
+        return list(self._timeframes.keys())
 
-    def all(self):
-        datasets = {}
-        for result in self.results:
-            datasets.update(result.all())
-        return datasets
+    def by_timeframes(self):
+        return self._timeframes
 
-    def keys(self):
-        keys = []
-        for result in self.results:
-            keys += result.keys()
-        return keys
-
-    def get_catkeys(self, catnum):
-        ret = set()
-        for key in self.keys():
-            elems = key.split('/')
-            ret.add(elems[catnum])
-        return list(ret)
+    def __getitem__(self, key):
+        if key in self._datasets:
+            return self._datasets[key]
+        elif key in self._timeframes:
+            return self._timeframes[key]
+        elif key in self._symbols:
+            return self._symbols[key]
+        raise KeyError('unrecognized TimeBucketKey, timeframe, or symbol: {}'
+                       ''.format(key))
 
-    def symbols(self):
-        return self.get_catkeys(0)
+    def __contains__(self, key):
+        return (key in self._datasets
+                or key in self._timeframes
+                or key in self._symbols)
 
-    def timeframes(self):
-        return self.get_catkeys(1)
+    def __iter__(self):
+        return iter(self._datasets.values())
 
-    def by_symbols(self):
-        datasets = self.all()
-        ret = {}
-        for key, dataset in six.iteritems(datasets):
-            symbol = key.split('/')[0]
-            ret[symbol] = dataset
-        return ret
+    def __str__(self):
```

Comment:
This is a total nit-pick, but I like to try to follow the convention of  `instance == exec(instance.__repr__())`, implementing `__str__` when something different is wanted for display purposes